torchao/quantization/README.md

```diff
@@ -0,0 +1,114 @@
+
+## Benchmarks
+Benchmarks are run on a machine with a single A100 GPU in `torchtune`.
+
+| Model       | Technique          | wikitext-perplexity | Tokens/Second | Memory Bandwidth (GB/s) |
+| ----------- | ------------------ | ------------------- | ------------- | ----------------------- |
+| Llama-2-7B  | Base (bfloat16)    | 8.789390849382297   |  20.16        | 316.26                  |
+|             | 8-bit              | 8.788388896424118   |  27.81        | 251.81                  |
+|             | 4-bit (G=256)      | 9.618806853408442   |  64.23        | 375.77                  |
+|             | 4-bit GPTQ (G=256) | 9.1791455391884     |  64.81        | 379.19                  |
+
+## Autoquantization
+
+The `autoquant` api can be used to quickly and accurately quantize your model. When used as in the example below, the api first identifies the shapes
+of the activations that the different linear layers see, it then benchmarks these shapes across different types of quantized and non-quantized layers in order to pick the fastest one, attempting to take into account fusions where possible. Finally once the best class is found for each layer, it swaps the linear. Currently this api chooses between no quantization, int8 dynamic quantization and int8 weight only quantization for each layer.
+
+```python
+import torch
+import torchao
+
+# inductor settings which improve torch.compile performance for quantized modules
+torch._inductor.config.force_fuse_int_mm_with_mul = True
+torch._inductor.config.use_mixed_mm = True
+
+# Plug in your model and example input
+model = torch.nn.Sequential(torch.nn.Linear(32, 64)).cuda().to(torch.bfloat16)
+input = torch.randn(32,32, dtype=torch.bfloat16, device='cuda')
+
+# perform autoquantization
+torchao.autoquant(model, (input))
+
+# compile the model to improve performance
+model = torch.compile(model, mode='max-autotune')
+model(input)
+```
+
+
+## A8W8 Dynamic Quantization
+
+```python
+# Fuse the int8*int8 -> int32 matmul and subsequent mul op avoiding materialization of the int32 intermediary tensor
+torch._inductor.config.force_fuse_int_mm_with_mul = True
+from torchao.quantization import quant_api
+# convert linear modules to quantized tensor subclasses
+quant_api.change_linear_weights_to_int8_dqtensors(model)
+```
+
+## A16W8 WeightOnly Quantization
+
+```python
+from torchao.quantization import quant_api
+quant_api.change_linear_weights_to_int8_woqtensors(model)
+```
+
+This technique works best when the torch._inductor.config.use_mixed_mm option is enabled. This avoids dequantizing the weight tensor before the matmul, instead fusing the dequantization into the matmul, thereby avoiding materialization of a large floating point weight tensor.
+
+
+## A16W4 WeightOnly Quantization
```

Comment:
is it also possible to add an example on how to use GPTQ?

Reply:
yeah sure